app/jobs/subscription_placement_job.rb

```diff
@@ -4,9 +4,10 @@
 
 class SubscriptionPlacementJob < ActiveJob::Base
   def perform
-    ids = proxy_orders.pluck(:id)
-    ProxyOrder.where(id: ids).each do |proxy_order|
-      place_order_for(proxy_order)
+    ProxyOrder.transaction do
+      while proxy_order = proxy_orders.lock.first do
```

Comment:
If we're wrapping the entire process in a transaction and any kind of error is encountered here, the entire thing will get rolled back, right? Every proxy order that had been (theoretically) processed will be reset to the state it was in when the transaction started. But within that we're also creating orders, line_items, payments, all kinds of stuff. It also means we'll be keeping a huge meta-transaction (which contains all the other transactions) open for the entire length of the processing (up to 5-6 minutes?).

But... we're also writing log entries and updating the details held in the `summarizer` object at various points in that process, and those changes _won't_ be rolled back, right? (they're effectively outside of the transaction)

I'm a bit nervous of the potential consequences here if we hit some unhappy paths...

Reply:
> If we're wrapping the entire process in a transaction and any kind of error is encountered here, the entire thing will get rolled back, right?

Yeah, that's a really good point. I'm now not sure how to resolve that though because we need the transaction for locking. 

If we want to lock all proxy orders in one batch then we need the big transaction around everything. Not great.

If we want smaller transactions, we need to do it per proxy order. So we either query each proxy order individually, as I'm doing at the moment or we query them all at once, without lock, and then lock them individually which is an additional query per proxy order.

We can also use the old approach of querying all the ids first and then loading them with a lock one-by-one within a smaller transaction. But since we then the proxy order selection is not synchronised and within once we acquired the lock on each we need to double check that it hasn't been processed yet. That's my best idea so far.

Reply:
Since transactions should be as short as possible, I moved the locking into the service class which is actually doing the order placement. I think it's much better there. Please have another look.

Reply:
That looks much nicer :100: